### PR TITLE
CI Test work

### LINKED ISSUE #946: Run automated tests with Travis
Useful to see whether new PRs passes the tests or not.

@colineberhardt would you be able to do this?

### PR DESCRIPTION
**Pull request by samccone**
Master still runs deploy but all other branches and PR's will run the tests on the changes.

:sparkles: :dancer: :sparkles: 

I anticipate a bit of friction when we flip this on.. just with the sauce keys and such but, everything works on my fork :+1: 

* First we add the real remote and fetch the contents
* then we diff the current branch from masters contents
* then we get the examples folders.. for instance

examples/backbone/....
examples/angular/.....

then I awk out the middle name, which happens to be used with our test
suite to run each test. From there I use xargs to invoke the tests
running over a subset of all of the tests.

```shell
npm run test -- --framework=backbone --framework=angular
--framework=dogescript
```

---

commit 1 allows us to do this

`npm run test -- --framework=jquery --framework=backbone --grep 'should trim entered text'`

------

Fixes #946

### CODE REVIEW

**Comment by ColinEberhardt** [comment]
Wow, thanks @samccone - that is absolutely awesome, very happy to see this implemented :-)

One area that I think will cause friction is the implementations that do not fully adhere to the HTML specification. This is why added a laxMode concept:

https://github.com/tastejs/todomvc/tree/master/browser-tests#lax-mode

The frameworks that require the laxMode setting isn't documented anywhere. Perhaps this should be recorded within `allTests.js` alongside the other metadata.

I think the tests code needs a bit of a tidy-up, but having them as part of the CI should ensure they are run and maintained. I was hoping to find some time to give them a bit of attention in the near future.

Anyhow, bug thumbs-up from me :+1:

**Comment by addyosmani** [comment]
> The frameworks that require the laxMode setting isn't documented anywhere. Perhaps this should be recorded within allTests.js alongside the other metadata.

:+1: 

> Anyhow, bug thumbs-up from me

Same! This is terrific work.

File: browser-tests/test.js [hunk 1]
```diff
@@ -21,9 +21,26 @@ module.exports.todoMVCTest = function (frameworkName, baseUrl, speedMode, laxMod
 		}
 
 		function launchBrowser() {
-			browser = new webdriver.Builder()
-			.withCapabilities({browserName : browserName})
-			.build();
+
+			if (process.env.SAUCE_USERNAME != undefined) {
+				browser = new webdriver.Builder()
+				.usingServer('http://'+ process.env.SAUCE_USERNAME+':'+process.env.SAUCE_ACCESS_KEY
```
Comment (arthurvr) [inline browser-tests/test.js]: Spaces around operators.

File: browser-tests/test.js [hunk 2]
```diff
@@ -21,9 +21,26 @@ module.exports.todoMVCTest = function (frameworkName, baseUrl, speedMode, laxMod
 		}
 
 		function launchBrowser() {
-			browser = new webdriver.Builder()
-			.withCapabilities({browserName : browserName})
-			.build();
+
+			if (process.env.SAUCE_USERNAME != undefined) {
```
Comment (arthurvr) [inline browser-tests/test.js]: Jshint complaint in here :D Use triple equal.

File: browser-tests/test.js [hunk 3]
```diff
@@ -21,9 +21,26 @@ module.exports.todoMVCTest = function (frameworkName, baseUrl, speedMode, laxMod
...
+					accessKey: process.env.SAUCE_ACCESS_KEY
+				})
+				.build();
+			} else {
+				browser = new webdriver.Builder()
+				.withCapabilities({
+					browserName : browserName
+				})
+				.build();
+			}
```
Comment (arthurvr) [inline browser-tests/test.js]: Actually there are 9 jscs errors on this piece of js

**Comment by arthurvr** [comment]
Great work Sam!

File: .travis.yml [hunk 4]
```diff
@@ -8,11 +8,12 @@ node_js:
...
     irc: 'irc.freenode.org#tastejs'
-branches:
-    only:
-        - master
+addons:
+    sauce_connect: true
 env:
     global:
+        - secure: K1WJRK2WTcc8+ZVGfwgOR8dBSM3VqJXNzcFpRPbEd48kVikx6nKYhrSffyl8a5QM5+Z7UP1TrmRAmOwR09Qp7Lh2Ch7vE9ooK9gk3XsoUP454aWm35NzjqagyajzGe7disCf6nWxgPZ0lgNkCG18aBfyh9FlUryzGzRyiSqeVeQ=
+        - secure: Or0NreVqcSIE6lbmSe3TmeXlm/++hXZAN9LVozvwlRLrE2jevIYRWD6arep9RTACtiQOPLrLrMh1/KDlIMgSfdrSa/w72J+YchSrYm+d06p4BcHPZEDnooiktTx4e9nL/5MfxIUYNOF0iT+wReRxId0INukqIUvjKB4F/X7MgmE=
```
Comment (arthurvr) [inline .travis.yml]: People might wonder what is included in these. Might be worth it to include some comments similar to the ones on lines 16 till 27.

File: browser-tests/test.js (hunk 3 above)
Comment (samccone) [inline browser-tests/test.js]: i only had 2... fixed though

File: browser-tests/allTests.js [hunk 5]
```diff
@@ -63,7 +63,9 @@ list = list.filter(function (framework) {
 // if a specific framework has been named, just run this one
 if (argv.framework) {
 	list = list.filter(function (framework) {
-		return framework.name === argv.framework;
+		return [].concat(argv.framework).some(function(f) {
```
Comment (arthurvr) [inline browser-tests/allTests.js]: `function(f) {` -> `function (f) {`

Comment (samccone) [inline browser-tests/allTests.js]: kk

Comment (arthurvr) [inline browser-tests/allTests.js]: ;)

**Comment by samccone** [comment]
updated :sparkles:

File: travis-runner.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  cp ../dist/.* .
+  git add -f .
+  git config user.email 'travis@rdrei.net'
+  git config user.name 'TasteBot'
+  git commit -am 'update the build files for gh-pages [ci skip]'
+  # Any command that using GH_OAUTH_TOKEN must pipe the output to /dev/null to not expose your oauth token
+  git push https://${GH_OAUTH_TOKEN}@github.com/${GH_OWNER}/${GH_PROJECT_NAME} HEAD:gh-pages > /dev/null 2>&1
+else
+  git remote add real https://github.com/tastejs/todomvc.git && \
+  git fetch real && \
```
Comment (arthurvr) [inline travis-runner.sh]: Real seems like a confusing name.

Comment (samccone) [inline travis-runner.sh]: thoughts on a better name?

Comment (arthurvr) [inline travis-runner.sh]: Just some brain farts but what about `current` or `stable` or so

Comment (samccone) [inline travis-runner.sh]: sure current is fine, will update

File: browser-tests/test.js [hunk 7]
```diff
@@ -21,9 +21,30 @@ module.exports.todoMVCTest = function (frameworkName, baseUrl, speedMode, laxMod
...
 
 		function launchBrowser() {
-			browser = new webdriver.Builder()
-			.withCapabilities({browserName : browserName})
-			.build();
+
+			if (process.env.SAUCE_USERNAME !== undefined) {
+				browser = new webdriver.Builder()
+				.usingServer(
+						'http://' +
```
Comment (arthurvr) [inline browser-tests/test.js]: Two tabs seems weird over here.

Comment (samccone) [inline browser-tests/test.js]: i agree, on it

**Comment by samccone** [comment]
did anyone else want to take a look at this?

File: browser-tests/test.js (hunk 7 above)
Comment (arthurvr) [inline browser-tests/test.js]: now it's totally not indented. That's even more weird. You just need one tab.

Comment (samccone) [inline browser-tests/test.js]: :+1: fixed

**Comment by samccone** [comment]
@arthurvr your style comments have been addressed :)

**Comment by arthurvr** [comment]
If nobody comlains let's land this one :+1:

**Pull request merged**